Add a wrapper for inspect in JIT to produce better error message

dzhulgakov opened the pull request:
Summary: If source code is not available due to packaging (e.g. sources are compiled to .pyc), TorchScript produces very obscure error message. This tries to make it nicer and allow to customize message by overriding _utils_internal.

Test Plan: Really hard to unittest properly. Did one off testing by compiling to .pyc and checking the message.

Differential Revision: D17118238

driazati commented on torch/_utils_internal.py:
```diff
@@ -33,5 +34,23 @@ def resolve_library_path(path):
     return os.path.realpath(path)
 
 
+def get_source_lines_and_file(obj):
```
This should probably live in `_jit_internal.py` instead

dzhulgakov replied:
_utils_internal.py is a magical file that can be overloaded in the build system with better error messages or internal-specific tweaks (see the comment). So that's why it lives in _utils_internal.

driazati commented on torch/_utils_internal.py:
```diff
@@ -33,5 +34,23 @@ def resolve_library_path(path):
+    except OSError as e:
+        raise OSError((
+            "Can't get source for {}. TorchScript requires source access in order to carry out compilation. " +
+            "Make sure original .py files are available. Original error: {}").format(filename, e))
```
Won't `filename` usually be `None` when this error is thrown? Maybe it should print the `obj` as well

driazati commented on torch/_utils_internal.py:
```diff
@@ -33,5 +34,23 @@ def resolve_library_path(path):
+    """
+    filename = None  # in case getsourcefile throws
+    try:
+        filename = inspect.getsourcefile(obj)
```
It's rare but I've run into it before, this can throw a `TypeError` if `obj` is a built-in, https://docs.python.org/3/library/inspect.html#inspect.getsourcefile, we should catch that too

dzhulgakov replied:
I think that's fine - I only want to enhance error message if source is not available. I guess if it's a builtin - it'd be either a problem with user input explicitly (like `torch.jit.script(open)`) or a bug in jit. Default message is pretty ok (`TypeError: module, class, method, function, traceback, frame, or code object was expected, got builtin_function_or_method`)

Commit:
Add a wrapper for inspect in JIT to produce better error message (#25415)

Summary:
If source code is not available due to packaging (e.g. sources are compiled to .pyc), TorchScript produces very obscure error message. This tries to make it nicer and allow to customize message by overriding _utils_internal.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/25415

Test Plan: Really hard to unittest properly. Did one off testing by compiling to .pyc and checking the message.

Differential Revision: D17118238

Pulled By: dzhulgakov

fbshipit-source-id: 4e72afe7898ea6ffc113ec5e4e55e8a859cb8f01
